docs/HowToFAQ/FAQReadsClause.md

```diff
@@ -0,0 +1,22 @@
+---
+title: Where do I put the reads clause?
+---
+
+## Question:
+
+This example
+```dafny
+{% include_relative FAQReadsClause.dfy %}
+```
+generates this error:
+```text
+{% include_relative FAQReadsClause.txt %}
+```
+but there is no obvious place to put a `reads` clause.
+
+## Answer:
+
+There is no place for the `reads` clause because no such clause should be needed.
+A type definition is not allowed to depend on a mutable field;
+the predicate in the subset type must be pure.
```

Comment:
Can you also explain a workaround?

You could say something along the lines that the idiom is to add a predicate Valid() on the class that reads its fields, and to ensure the value conforms to the predicate, you can put `Valid()` in pre and post conditions of methods, in pre-conditions of functions and predicates.

```dafny
class A {
  var x: string
  var y: string
  predicate Valid() reads this {
    |x| == |y|
  }
}

method Test(a: A)
  requires a.Valid()
  requires a.x == ""
  modifies a
  ensures a.Valid()
{
  assert a.Valid(); // That's correct thanks to the precondition.
  a.x := "abc"; // Because |a.y| == 0, we broke the Valid() predicate
  assert !a.Valid(); // But that's ok.
  a.y := "bca";
  assert a.Valid(); // Now Dafny can prove this
  // The postcondition holds
}
```

Can you please also link to the `{:autocontracts}` attribute?

Reply:
Done